### ISSUE TITLE
pm: Implement orchestrator strategy for generating ticket parameters

### ISSUE THREAD

**Issue by yondonfu** [livepeer/go-livepeer#828]

**Comment by yondonfu** [comment]
O could generate ticket parameters based on B's reserve. An alternative strategy is to use a fixed faceValue & winProb with B varying the frequency of ticket emission (i.e. sending N tickets per segment depending on the transcoding price). Additionally, @josh states the following:

> This is why it feels like having a mostly fixed faceValue / winProb is a more robust approach. The impact on the network seems easier to reason about, in addition to fixing the overhead for B/O. Eg, we can approximate the overall frequency of winning tickets to something like "once per X hours of 720p video transcoded," but this becomes more involved if faceValue / winProb are dynamic based on a given B's reserve.

My questions regarding to using a fixed faceValue & winProb are:

- Wouldn't you still need to adjust faceValue & winProb whenever the gas price changes to maintain a fixed overhead?
- Wouldn't the overall frequency of winning tickets ultimately depend on the frequency of ticket emission which will vary depending on the transcoding pricing on the network (i.e. O_1's transcoding pricing results in 5 tickets per segment while O_2's transcoding pricing results in 2 tickets per segment)?

If the answer to these questions is yes, would using a fixed faceValue & winProb really make the impact on network congestion (in terms of frequency of on-chain ticket redemption) easier to reason about?

**Comment by j0sh** [comment]
> Wouldn't you still need to adjust faceValue & winProb whenever the gas price changes to maintain a fixed overhead?

In retrospect, the use of the word 'fixed' was clumsy. The idea was that the faceValue / winProb would be *common* (or shared) among all Bs using the O, rather than deriving different values based on each B's reserve.

In the case where the gas price fluctuates then you are correct that the faceValue / winProb would have to adjust to maintain the same amount of overhead. That being said, there are a few ways to accommodate fluctuations in gas price:

* Adjust only the faceValue and have B pay for the difference in overhead. Frequency of redemption is unchanged and this wouldn't really help immediate congestion, but ensures the transaction will go through and maintains O's margins.

* Adjust both faceValue / winProb to maintain the overhead ratio. Tickets will come in more (or less) frequently, which may (or not) help with congestion.

Can we make the same types of adjustments if the faceValue is derived from the float on a per-B basis? It seems there would be less leeway for adjustment if the numbers were tight, eg a small reserve.

* O swallows any difference within a reasonable range; this mitigates the need to re-issue PM parameters every time gas fluctuates a few gwei. Although at 1-2gwei safelow, that could easily mean a doubling of the overhead, however small. Frequency of redemption is unchanged.

> Wouldn't the overall frequency of winning tickets ultimately depend on the frequency of ticket emission which will vary depending on the transcoding pricing on the network

Yes, the winning frequency per unit of *time* would depend on how frequently tickets are emitted. Emission in turn depends on the transcoding configuration that B wants, rather than the per-unit price of transcoding.

The winning ticket frequency would still be constant per unit of *work*. If winning tickets were being generated too frequently in time for a given transcoding configuration (say, sub-1 hour), then the PM parameters can be adjusted to a more comfortable level for that particular stream.

It might help to think of the faceValue in terms of upper bounds for transcoding work. Suppose one ticket nominally covers two hours of 720p transcoding (regardless of price).  If per-unit pricing is constant for an O and B actually wants to transcode 720p, 480p and 360p that ends up covering roughly 1.2 hours of work : `2 * 1280*720 / (1280*720 + 858*480 + 480*360) = 1.22`

> If the answer to these questions is yes, would using a ~~fixed~~ **common** faceValue & winProb really make the impact on network congestion (in terms of frequency of on-chain ticket redemption) easier to reason about?

(changed 'fixed' to 'common') I do think so. Bs using a given O would be sharing mostly the same parameters, making the calculations straightforward on a per-O basis.

**Comment by j0sh** [comment]
Continuing the discussion from here: https://github.com/livepeer/wiki/pull/36#issuecomment-483869548

> What if you default to guaranteeing 2 winning tickets and if the reserve is too small such that it would result in ticket redemptions that would be too frequent (i.e. too low faceValue and too high winProb) then O would not work with B?

That's certainly an option. Seems that O has a few choices here for small [1] reserve amounts:
* Guarantee itself a few winning tickets, regardless of overhead or frequency. Potentially get a smaller cut of each ticket due to gas overhead being larger.
* Guarantee itself a single, larger winning ticket covering the full amount of the float.
* Be a polite network citizen and decline to work with a B if O deems the reserve too small given prevailing network conditions.

To me, the second option seems to be the best compromise. The O is getting *something* while minimizing the overhead for B and the impact on the network.

It does feel that the ultimate decision on whether to accept a given amount of PM overhead should rest on B, rather than O, since B is the one that is shouldering the cost. Externalities such as network congestion still come back to B in terms of higher overhead or faceValues, so B still has an interest in minimizing its overhead. If not, B would not be able to get any work done for smaller reserve amounts.

[1] "small" reserve amounts being an amount smaller than the O's default faceValue, if we do end up assigning a default faceValue / winProb for all Bs with a given O.

**Comment by yondonfu** [comment]
> Can we make the same types of adjustments if the faceValue is derived from the float on a per-B basis? 

I think we can via the following formula for `winProb`:

```
winProb = price / (faceValue - txCost)

# This is derived from the following formula for faceValue
# Calculating faceValue in this way would always price in txCost when generating
# ticket parameters. As a result, we can guarantee that O is always compensated
# for txCost in addition to the transcoding price it is charging
faceValue = (price / winProb) + txCost
```

Given a `faceValue` based on B's max float (we can say that the `faceValue` is the max float for simplicity), if the gas price increases then `winProb` increases. Despite `faceValue` remaining the same, the EV of the ticket is now higher which reflects B paying additional overhead on top of the base transcoding price.

O wouldn't care about increased transaction overhead because the cost is always passed on to B and B is the one that shoulders that burden. If we calculate `faceValue` based on B's max float, then Bs with smaller reserves will pay higher additional overhead on top of the base transcoding price. As long as Bs can define their own acceptable pricing (i.e. a maximum) then I think the node could always alert the user that the current reserve amount is too small as it results in Os requiring ticket parameters corresponding to prices that are too high for B.

> O swallows any difference within a reasonable range

I think this is reasonable especially if O uses a gas price that is in the fast or standard range instead of the safe low range. Then, if the average gas price increases such that the old standard range is now the new safe low range, O could still redeem the ticket for the same profit it will just take longer for the transaction to confirm on-chain.

> Yes, the winning frequency per unit of time would depend on how frequently tickets are emitted. 

If we care about determining the frequency of winning tickets per unit of time in the interest of limiting potential network congestion i.e. by targeting 1 win per hour, wouldn't sending a ticket per segment with dynamically set `faceValue` and `winProb` based on the # of pixels in a segment, B's reserve and current gas prices instead of using a common `faceValue` and `winProb` make determining the frequency of winning tickets per unit of time easier? You have different `winProb` values across streams, but at least you know you have 1 ticket every 2 seconds per stream.

> Bs using a given O would be sharing mostly the same parameters, making the calculations straightforward on a per-O basis.

Wouldn't they not be using mostly the same parameters if O uses one of the adjustment strategies you mentioned when gas price fluctuates or if the common `faceValue` is greater than a B's max float?

> To me, the second option seems to be the best compromise. The O is getting something while minimizing the overhead for B and the impact on the network.

Right - with solid failover having O just set face value to match the max float feels like it could be reasonable since B would just switch to a different O if the current O receives a winning ticket and stops accepting tickets until it redeems the winning ticket.

**Comment by yondonfu** [comment]
While considering setting `faceValue` & `winProb` based on B's reserve, here is another thought:

- O sets `faceValue` to the `maxFloat` derived from B's reserve
- O sets `baseWinProb` so that the EV of the ticket pays for a single pixel. Thus, `baseWinProb = pricePerPixel / (faceValue - txCost)`
- When B sends a ticket for a segment, it computes the # of pixels in the segment as `pixelCount`. Then, it computes the `winProb` for the ticket as `winProb = pixelCount * baseWinProb` (which is equivalent to multiplying `pricePerPixel` in the numerator of `baseWinProb` by `pixelCount` in order to determine the price for `pixelCount` # of pixels)
- When O receives a ticket it counts the # of pixels in a segment and checks that the ticket has an acceptable `faceValue` and also that the `winProb` of the ticket is the result of multiplying the required `baseWinProb` by the # of pixels in the segment

**Comment by j0sh** [comment]
Interesting, so the winProb becomes dynamic based on the number of pixels in the segment. Some thoughts about that:

* Does this require that `faceValue` be set based on B's reserve, or could the `faceValue` still be common among all B using the O?

* How could we incorporate various types of accounting into the `baseWinProb` ? In addition to per-pixel accounting, suppose there is per-sample (audio) accounting, or accounting for decoding costs (4k@120fps AV1 decode takes a lot more resources than 720p@30fps H.264), and so forth. Also, depending on how we do pricing, the pricePerPixel may not necessarily be a constant (different prices for different codecs). How would these be reconciled to generate the final winProb?

* Will this require a "pre-processing" step to check the segment to ensure that the winProb is approximately correct? If there is no such pre-processing step, how should we handle tickets that are unacceptable for the segment (eg, too low winProb)? Should we be short-circuiting the transcode after a certain period? This may require deeper integration with the transcoder.

* We likely will need to synchronize the accounting state, as B's initial calculation of the number of output pixels will be an approximation (different implementations between B/O may lead to slightly different results). The under / over can be communicated back after each segment is transcoded. To accommodate this, subsequent ticket's winProb could be adjusted based on the difference but that feels more difficult to reason about. When the winProb/faceValue are 'fixed' then B just needs to send more (or fewer) tickets to compensate for the accounting discrepancy. Naturally we'll need some reasonable limit here not to take advantage of any drift; this probably will take the form of B and O both sanity checking the numbers in the transcoded results. These are all accounting particulars that we don't need to settle immediately, but just wanted to point out some potential knock-on consequences for PM.

**Comment by yondonfu** [comment]
> Does this require that faceValue be set based on B's reserve, or could the faceValue still be common among all B using the O?

Not necessarily, but if a common `faceValue` is used among all Bs using the O there seems to be two options:

1) If O needs to have the `faceValue` guaranteed by B's reserve then O adjusts the `faceValue` down for Bs with smaller reserves which would be similar to always setting `faceValue` based on B's reserve
2) If O doesn't need to have the `faceValue` guaranteed (perhaps because it believes that B has a sufficient deposit such that the remaining amount of the `faceValue` not covered by B's reserve will be available in B's deposit by the time O's ticket redemption confirms on-chain), then O could always use the same `faceValue` across all Bs

Option 1 seems simpler since O doesn't need to guess at whether B's deposit is sufficient.

> How could we incorporate various types of accounting into the baseWinProb ?

Could decoding costs be incorporated into the price per pixel such that the price reflects the cost of decoding the pixel and then encoding the pixel based on all of the requested transcoding options?

For audio accounting, could a separate set of ticket parameters be issued for a single sample based on a price per sample? Then, for each video segment, B would send 1 ticket that pays for the # of pixels included in the segment and 1 ticket that pays for the # of samples included in the segment.

> Will this require a "pre-processing" step to check the segment to ensure that the winProb is approximately correct?

I imagined the pre-processing step to have O count the pixels of segment and then multiply the pixel count by O's current required `baseWinProb` to check if the `winProb` of the ticket that came with the segment is valid. But, I think I see what you're getting at...the pixel counting functionality resides in the transcoder so if we wanted to do this the orchestrator would need to be more tightly integrated with the transcoder so that it can either have the transcoder count the pixels of a segment or if the orchestrator is bundled with the transcoder then it can count the pixels of a segment itself? 

> as B's initial calculation of the number of output pixels will be an approximation (different implementations between B/O may lead to slightly different results

Does B need to approximate the output pixels for a given input pixel and a set of transcoding options or could O just price that into the price per pixel i.e. O says "this is the cost of transcoding your single input pixel for this set of transcoding options and I have already estimated the # of output pixels produced from your input pixel for each of the transcoding options"?

**Comment by j0sh** [comment]
> Option 1 seems simpler since O doesn't need to guess at whether B's deposit is sufficient.

Agreed. FWIW, I was thinking something like this:

```
faceValue = min(commonFaceValue, B_float)
winProb = faceValue / defaultFaceValue * commonWinProb
```

> the orchestrator would need to be more tightly integrated with the transcoder so that it can either have the transcoder count the pixels of a segment or if the orchestrator is bundled with the transcoder then it can count the pixels of a segment itself?

More broadly, one of the advantages of a looser coupling between payments and the work being done is that pre-processing is not necessary to validate the payment itself.

> O says "this is the cost of transcoding your single input pixel for this set of transcoding options and I have already estimated the # of output pixels produced from your input pixel for each of the transcoding options"?

I'm not sure; this seems to imply that the workflow is B sends the ticket *after* the transcoding is done. I was thinking the opposite: B sends the ticket up front along with the segment, and O doesn't necessarily need the ticket to match exactly with the amount of work that's being requested of the segment. Eg:

* B sends segment and ticket.
  * Presumably all tickets have an uniform EV for a given B
* O credits B based on value of ticket
* O processes segment
* O debits B based on value of work actually done
* O responds with to B with accounting update
* B sends segment, and another ticket if necessary.

**Comment by yondonfu** [comment]
After a few very helpful discussions with @j0sh I am now leaning toward a "fixed EV" approach for generating ticket params. While this approach will have O default to a particular set of ticket params to be used with all Bs, O might have to adjust the ticket params based on B's maxFloat while maintaining a fixed EV so I choose to refer to this as the "fixed EV" approach as opposed to the "common params" approach.

Before we discuss the fixed EV approach, let's examine an alternative: O could generate ticket params such that B always sends 1 ticket per segment which will pay O based on the # of output pixels in the transcoded results. Some of the challenges of this approach include:

- The goal of having 1 ticket pay for the # of pixels in the transcoded results means that we have to tightly couple the ticket params generation strategy with the details of payment accounting. As a result, any change in the details of payment accounting requires a change in the ticket params generation strategy. For example, if we choose to factor in input decoding or audio encoding in the future, the ticket params generation strategy would need to be modified such that ticket(s) pay for the # of pixels in the input, # of pixels in the transcoded results and # of samples (using samples here for convenience - this can be replaced with the actual base unit of work for audio encoding) in the encoded audio results. A clear separation between the ticket params generation strategy and payment accounting would allow for payment accounting to be updated in the future without any necessary changes in how ticket params are generated.
- If the goal is for 1 ticket to pay for an *exact* # of pixels in the transcoded results, then O needs to check that the actual # of pixels in the transcoded results is covered by a received ticket. O has two options:
  - Pre-processing: O can estimate the # of pixels in the transcoded results upon receiving an input segment and refuse to process the segment if the ticket does not pay for the estimated # of pixels. We should assume that B & O could be using different implementations which means that B's estimate for the # of pixels in the transcoded results for an input segment (which is calculated in order to set the params of a ticket) could be different from O's estimate. If B's estimate is different from O's estimate, then the ticket may overpay or underpay O. Another downside of pre-processing is that it adds an additional computational step prior to starting to transcoding a segment which can hurt latency.
  - Post-processing: O can process a segment, observe the exact # of pixels in the transcoded results and then take some action if the ticket did not pay for the exact # of pixels. This would have less of a latency impact than pre-processing, but the exact # of pixels in the transcoded results could still be different from B's estimate
  - Regardless of whether O uses a pre-processing or post-processing step, it seems that it is very possible that B's ticket will not pay for the exact # of pixels in the transcoded results. This prompts the question: why even try to have a ticket pay for an exact # of pixels in the transcoded results at all?

We can address the second point with a [credit/debit balance sheet based payment accounting strategy](https://github.com/livepeer/go-livepeer/issues/828#issuecomment-484957111) which is completely decoupled from how we choose to generate ticket params. Additional details of this payment accounting strategy are left for a separate GH issue.

We can address the first point with the fixed EV approach. The fixed EV approach involves some default settings for O and a `GenerateTicketParams()` algorithm:

- O configures `defaultOverhead` (i.e. 1%)
- O configures a desired EV which indicates that it wants to get paid for every EV amount of transcoding work it performs (i.e. .1 ETH would indicate that it wants to get paid for every .1 ETH amount of transcoding work it performs)
- `GenerateTicketParams()`
  - `currOverhead = defaultOverhead`
  - `faceValue = txCost / currOverhead`
  - If `faceValue > B.maxFloat`
    - `faceValue = B.maxFloat`
  - If `faceValue < EV` (would result in `winProb > 1`!)
    - return error
  - `winProb = EV / faceValue`
  - return `faceValue` and `winProb`

By default O will use a common set of ticket params that is derived from its desired default overhead (i.e. the % overhead relative to the `faceValue` of a ticket incurred by ETH transaction costs). If the common `faceValue` is greater than the `maxFloat` available for a B, O can adjust the `faceValue` to match B's `maxFloat` and also adjust `winProb` to maintain O's desired fixed EV.

While the the ticket params generation strategy mentioned above could be implemented without touching payment accounting, it is worth noting that O can also dynamically adjust its transcoding pricing to factor in the additional transaction costs incurred from ticket redemption:

- O configures a `basePricePerPixel` for each preset (for simplicity assume that pricing is just based on presets for now). This price does NOT attempt to take into account ETH transaction costs
- `GetPrice(preset)`
  - `currOverhead = defaultOverhead`
  - `faceValue = txCost / currOverhead`
  - if `faceValue > B.maxFloat`
    - `faceValue = B.maxFloat`
    - `currOverhead = txCost / faceValue`
  - `basePricePerPixel = getBasePricePerPixel(preset)`
  - `marginForOverhead = (currOverhead * basePricePerPixel) / (1 - currOverhead)`
  - `pricePerPixel = basePricePerPixel + marginForOverhead`
  - return `pricePerPixel`

By default O's price per pixel for each preset will reflect its default overhead (i.e. the price might reflect a 1% increase over its base price in order to reflect ETH transaction costs). If the common `faceValue` is greater than the `maxFloat` available for a B, O can adjust the `faceValue` to match B's `maxFloat`. This increases the actual % overhead represented by ETH transaction costs. O can factor this higher % overhead in its price per pixel for each preset.

B can determine how many tickets (each with a fixed EV) to send with each segment based on the value owed to O and an estimated # of pixels in the transcoded results for an input segment:

- `TicketsPerSegment(O, EV, seg, presets)`
  - `balanceSheet = getBalanceSheet()`
  - `value = balanceSheet.getValueOwed(O)`
  - `for preset in presets`
    - `outputPixels = estimateOutputPixels(seg, preset.id)`
    - `value += (outputPixels * preset.pricePerPixel)`
    - return `value / EV`
  
Note: The balance sheet related lines in `TicketsPerSegment()` assume the use of a credit/debit balance sheet payment accounting strategy - the tickets that B sends with a segment to O likely won't completely cover the value of the transcoding work so B might "owe" some value to O and there would be an expectation that B makes up for this drift with an additional ticket sent with the next segment.

**Comment by yondonfu** [comment]
With the latest [payment accounting proposal](https://github.com/livepeer/go-livepeer/issues/866), the relevant TODOs for this issue are:

Update `pm.Recipient` so that instead of being created with `faceValue` and `winProb` it is created with:

- `ticketEV`: The desired ticket EV
- `txGas`: The tx gas of redeeming a ticket
- `targetTxCostOverhead`: The target overhead for tx cost relative to ticket face value

Update the `TicketParams(sender)` method for `pm.Recipient` based on the following pseudocode:

```
currOverhead = recipient.targetTxCostOverhead
// txCost will use the recipient.txGas and fetched gas price to determine tx cost
faceValue = txCost(recipient.txGas) / currOverhead
// MaxFloat returns the max float for a sender based on its current reserve
maxFloat = MaxFloat(sender)
// TicketEV returns the fixed ticket EV that the node is started with
ticketEV = recipient.ticketEV

if faceValue > maxFloat {
    // Assume that when recipient is created, there is validation
    // to make sure that ticketEV >= the default faceValue calculated
    // when using the txCost and recipient.targetTxCostOverhead
    // Thus, if we have to lower faceValue to match maxFloat and maxFloat
    // is less than ticketEV, this means that the sender's reserve is insufficient
    if maxFloat < ticketEV {
        return ERROR_INSUFFICIENT_RESERVE
    }

    faceValue = maxFloat
}

winProb = ticketEV / faceValue

return TicketParams{
    ...
    faceValue,
    winProb
}
```

**Comment by yondonfu** [comment]
Closed by #904
